### ISSUE TITLE
TypeScript - Missing type in ActionType

### ISSUE THREAD

**Issue by julianomqs** [plopjs/plop#180]
Hello,

I'm trying to use plop with TypeScript, and I'm getting the following error.

```
Type '{ type: string; path: string; templateFile: string; }' is not assignable to type 'ActionType'.
  Type '{ type: string; path: string; templateFile: string; }' is missing the following properties from type 'AppendActionConfig': pattern, unique, separator, template, and 3 more.ts(2322)
```

My `plopfile.ts`

```typescript
import { NodePlopAPI } from "plop";

export default function(plop: NodePlopAPI) {
    // controller generator
    plop.setGenerator("controller", {
        description: "application controller logic",
        prompts: [
            {
                type: "input",
                name: "name",
                message: "controller name please"
            }
        ],
        actions: [
            {
                type: "add",
                path: "src/{{name}}.js",
                templateFile: "plop-templates/controller.hbs"
            }
        ]
    });
}
```

Image

![image](https://user-images.githubusercontent.com/7409904/67123350-6f9eaf80-f1c6-11e9-8c7d-5a8ad4b0c182.png)

(GUESS) Apparently the `ActionType ` lacks the `AddActionConfig` interface:

```typescript
export type ActionType =
  | ActionConfig
  | AddManyActionConfig
  | ModifyActionConfig
  | AppendActionConfig
  | CustomActionFunction;
```

**Comment by amwmedia** [comment]
I'm not an experienced TypeScripter, summoning @crutchcorn @cspotcode and @tannerntannern.

Any thoughts here?

**Comment by cspotcode** [comment]
@julianomqs is exactly right.

We are describing actions using a "discriminated union."  It's a union of interfaces, where each one has a `type` field that's a different string constant.  This means that as soon as a coder writes `"type": "add"` in their editor, the discriminated union kicks in, TypeScript narrows the union down to the only compatible choice -- the interface corresponding to the `"add"` action -- and the rest of the properties of the object get tab-completion and docs tooltips.

https://github.com/plopjs/node-plop/blob/master/index.d.ts#L72-L78

**Comment by cspotcode** [comment]
Also!  There are some cool tricks that allow third-party libraries and to contribute declarations for their own, custom action types.  These contributed declarations will be pulled into the union automatically, so that coders will get a good experience writing those custom actions.

For example, a coder types `{"type": "add-http-endpoint", <nice tooltips here>` and it just works.

**Comment by amwmedia** [comment]
Any chance you'd have time to submit a PR for the change that is needed to get this working? I've taken a look locally and can't seem to get the error to clear.

**Comment by crutchcorn** [comment]
@cspotcode I'll let you have first-chance to take a look, but if nothing else, I can take a look

**Comment by julianomqs** [comment]
I was peeking in the index.d.ts and I noticed this:

```typescript
export interface AddActionConfig extends ActionConfig {
  type: "add";
  path: string;
  template: string;
  templateFile: string;
  skipIfExists: boolean;
}
```

Correct me if I'm wrong, but all those fields in this type are required, right?

But the `template` and the `skipIfExists` are not required/has default value respectively.

**Comment by crutchcorn** [comment]
I believe you're correct @julianomqs . Perhaps you might be the right person to open a PR to fix some of the typings issues that I'd introduced 😉 👍

**Comment by cspotcode** [comment]
Work's rough for me right now, so I can't write PRs.  I can probably review them.

**Comment by crutchcorn** [comment]
@julianomqs do you want to take a stab at a PR against https://github.com/plopjs/node-plop to fix the `index.d.ts` file with the things you're describing?

**Comment by julianomqs** [comment]
Yes, I wouldn't mind helping.

But I'm a beginner in TypeScript so it would be nice someone with more knowledge reviewing it.

It is just a matter of editing the index.d.ts, right? Or are there other steps involved?

**Comment by crutchcorn** [comment]
We all are at some point 😄 We'll be here to review once you have a PR opened (please ping me when you make a PR in the PR body)

It should just be editing the `index.d.ts` file and testing, no other steps 😁

**Comment by julianomqs** [comment]
Ok then, so that's what I will do:

- Adding the `AddActionConfig` in the `ActionType` type;
- Setting the optionals/default values in the action types in respect with the docs;
- Review the file and try to identify another situations.

I will ping you when it's done.

**Comment by amwmedia** [comment]
you'll just need to `npm link` node-plop into plop in order to test I think

**Comment by cspotcode** [comment]
An easy way to write tests for the declarations is to write an example plopfile `.ts` and run `tsc` with the `noEmit` flag set.  This will still spit out type errors, so it'll fail if the declarations are wrong.

Another great option is `tsd` (https://github.com/SamVerschueren/tsd) or even https://github.com/microsoft/dtslint

`tsd` is really simple: create a `index.test-d.ts` and run `tsd` on the command-line.

**Comment by crutchcorn** [comment]
@cspotcode while I agree we want to setup TypeScript tests, I don't think it's within the scope of this issue, I'd rather break that out to it's own issue on the `node-plop` repo.

Plus, the setup for `dtslint`(which is the official answer from MS) is somewhat daunting for newcomers (or at least was for me). I don't mind either contributing those type tests or working with someone else to help mentor them with how they can set it up either

**Comment by cspotcode** [comment]
@crutchcorn I don't recommend `dtslint`; use `tsd`.  Setup for that is nonexistent.  Or just run `tsc` directly.  Don't get hung up on setup for tests, and don't confuse having alternatives with needing to consider them.  Take the easy option, and it'll be simple.

I think if @julianomqs wants to make sure they don't break anything, they're gonna be testing manually in their editor anyway.  So they might as well commit that file to git.  That's basically all you need; run `tsc` against it and that's testing taken care of.

**Comment by amwmedia** [comment]
Let's get the current issue fixed, and then someone can add tests within `node-plop` to ensure everything is working as expected. 👍

**Comment by jedmao** [comment]
Here it is, y'all

https://github.com/plopjs/node-plop/pull/132

**Comment by amwmedia** [comment]
released in `plop@2.5.3`. @julianomqs please have a look and see if this solves the issue for you.

**Comment by FDiskas** [comment]
Getting same error on `plop@2.5.3` :heart_decoration:
